### PR TITLE
Initial implementation of Environs.Network Interface for OpenStack Pr…

### PR DESCRIPTION
**Pull request by hmlanigan**
…ovider.

Updates to goose required to complete.

## Please provide the following details to expedite Pull Request review:

----

## Description of change

This is part of the Cross Model Relations work and needed for juju Networking.

environs.Networking is implemented for the OpenStack Provider.  Neutron networking will be
supported here, though not Nova networking.  Updates will be made once more data is available from goose.

A small bug was fixed in supportsNeutron() as well to make Neutron networking the default if
not OpenStack Endpoints are found.

## QA steps

1. juju bootstrap with an openstack cloud (utilizing neutron networking)
2. create two models
3. in model one: 
         juju deploy mysql
         juju offer mysql:db hosted-db    <-- copy url from results
         juju show-endpoints
4. in model two:
        juju deploy ghost
        juju consume <url-for-hosted-db> 
        juju add-relation ghost:database hosted-db:db
5. validate results of juju list-subnets
6. the neutron security group for the machine mysql is running on should have a rule allowing traffic on port 3306 from the CIDR of the subnet the machine for ghost is running on.
7. both charms should report a good status

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

**Comment by hmlanigan** [comment]
!! testme !!

File: provider/openstack/provider.go [hunk 1]
```diff
@@ -648,6 +649,11 @@ func (e *Environ) Bootstrap(ctx environs.BootstrapContext, args environs.Bootstr
 func (e *Environ) supportsNeutron() bool {
 	client := e.client()
 	endpointMap := client.EndpointsForRegion(e.cloud.Region)
+	if len(endpointMap) == 0 {
+		// In some corner cases the endpoingMap is empty.  Default to
```
Comment (axw) [inline provider/openstack/provider.go]: I think you'll only get an empty map back if the client has not authenticated? Probably need to change the order of some code, so that auth happens first.

Comment (hmlanigan) [inline provider/openstack/provider.go]: the 3 places calling supportsNeutron() try to get an authenticated client if the client isn't already.... looks like something else is going on.  will investigate.  all the goose output of endpoint data i saw did include "network".

File: provider/openstack/provider.go [hunk 2]
```diff
@@ -1003,6 +1009,7 @@ func (e *Environ) StartInstance(args environs.StartInstanceParams) (*environs.St
 		return nil, errors.Annotate(err, "getting initial networks")
 	}
 	usingNetwork := e.ecfg().network()
+	logger.Debugf("ecfg().network() returned (%s)", usingNetwork)
```
Comment (wallyworld) [inline provider/openstack/provider.go]: probably don't need this to be committed

**Comment by hmlanigan** [comment]
!! testmeagain !!

File: provider/openstack/legacy_networking.go [hunk 3]
```diff
@@ -73,3 +75,13 @@ func (n *LegacyNovaNetworking) ResolveNetwork(name string) (string, error) {
 	}
 	return processResolveNetworkIds(name, networkIds)
 }
+
+// Subnets is part of the Networking interface.
+func (n *LegacyNovaNetworking) Subnets(instId instance.Id, subnetIds []network.Id) ([]network.SubnetInfo, error) {
+	return nil, errors.NotSupportedf("nova subnet")
```
Comment (wallyworld) [inline provider/openstack/legacy_networking.go]: If we are going to do this here, then the discover spaces worker will need to be tweaked to handle this error and ignore it, since the worker thinks these methods should work. Add a todo also to the worker to revert the change when nova subnet support is added.
Should also add a test for this case.

File: provider/openstack/networking.go [hunk 4]
```diff
@@ -246,3 +274,82 @@ func resolveNeutronNetwork(neutron *neutron.Client, name string) (string, error)
...
+	logger.Tracef("found subnet with info %#v", info)
+	return info, nil
+}
+
+// Subnets returns basic information about the specified subnets known
+// by the provider for the specified instance or list of ids. subnetIds can be
+// empty, in which case all known are returned.
+func (n *NeutronNetworking) Subnets(instId instance.Id, subnetIds []network.Id) ([]network.SubnetInfo, error) {
+	var results []network.SubnetInfo
+	subIdSet := make(map[string]bool)
```
Comment (wallyworld) [inline provider/openstack/networking.go]: we have a juju/utils/set package which has a strings set implementation which will simplify the code below

Comment (hmlanigan) [inline provider/openstack/networking.go]: thank you - updated the function.

File: provider/openstack/provider.go (hunk 1 above)
Comment (hmlanigan) [inline provider/openstack/provider.go]: It turns out that even though the callers to supportsNeutron() were authenticating the client before calling.  The client used by suportsNeutron() was not alway authenticated.  There is a bigger question of why that should be looked at?  I moved code around so that supportsNeutron() was checking for client authentication for now.

File: provider/openstack/networking.go [hunk 5]
```diff
@@ -246,3 +273,75 @@ func resolveNeutronNetwork(neutron *neutron.Client, name string) (string, error)
...
+		// TODO(hml): 2017-03-20
+		// Implment Subnets() for case where instId is specified
+		return nil, errors.NotImplementedf("neutron subnets with instance Id")
+	} else {
+		neutron := n.env.neutron()
+		subnets, err := neutron.ListSubnetsV2()
+		if err != nil {
+			return nil, errors.Annotatef(err, "failed to retrieve subnets")
+		}
+		if len(subnetIds) == 0 {
```
Comment (wallyworld) [inline provider/openstack/networking.go]: This logic can be simplified using the functionality provided by Set. You just need to do a set difference between the set of passed in ids and those obtained by the list subnets call.

Comment (hmlanigan) [inline provider/openstack/networking.go]: using Set

File: provider/openstack/legacy_networking.go (hunk 3 above)
Comment (hmlanigan) [inline provider/openstack/legacy_networking.go]: updated the worker.

File: worker/discoverspaces/discoverspaces.go [hunk 6]
```diff
@@ -154,11 +154,16 @@ func (dw *discoverspacesWorker) handleSubnets() error {
 }
 
 func (dw *discoverspacesWorker) discoverSubnetsOnly(environ environs.NetworkingEnviron) error {
-	modelSubnetIds, err := dw.getModelSubnets()
-	if err != nil {
+	subnets, err := environ.Subnets(instance.UnknownId, nil)
+	if err != nil && errors.IsNotSupported(err) {
```
Comment (wallyworld) [inline worker/discoverspaces/discoverspaces.go]: you don't need the err != nil

**Comment by hmlanigan** [comment]
!! runtest !!

**Comment by hmlanigan** [comment]
$$ merge $$

**Comment by anastasiamac** [comment]
I think you may need to have $$merge$$ without spaces...

**Comment by hmlanigan** [comment]
$$merge$$
